src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,553 @@
+        hook is always called after `on_train_epoch_end` and `on_validation_epoch_end`. It is
+        called regardless of these hooks' return values.
+
+        Parameters
+        ----------
+        epoch_no
+            The current epoch (the first epoch has `epoch_no = 0`).
+        epoch_loss
+            The validation loss that was recorded in the last epoch if validation data was
+            provided. The training loss otherwise.
+        training_network
+            The network that is being trained.
+        trainer
+            The trainer which is running the training.
+        best_epoch_info
+            Aggregate information about the best epoch. Contains keys `params_path`, `epoch_no` and
+            `score`. The score is the best validation loss if validation data is provided or the
+            best training loss otherwise.
+        ctx
+            The MXNet context used.
+
+        Returns
+        -------
+        bool
+            A boolean whether the training should continue. Defaults to `True`.
+        """
+        return True
+
+    def on_train_end(
+        self,
+        training_network: nn.HybridBlock,
+        temporary_dir: str,
+        ctx: mx.context.Context = None,
+    ) -> None:
+        """
+        Hook that is called after training is finished. This is the last hook to be called.
+
+        Parameters
+        ----------
+        training_network
+            The network that was trained.
+        temporary_dir
+            The directory where model parameters are logged throughout training.
+        ctx
+            An MXNet context used.
+        """
+
+
+class CallbackList(Callback):
+    """
+    Used to chain a list of callbacks to one Callback.
+    Boolean hook methods are logically joined with AND, meaning that if at least one callback method returns False, the training is stopped.
+
+    Parameters
+    ----------
+    callbacks
+        A list of gluonts.mx.trainer.callback.Callback's.
+    """
+
+    @validated()
+    def __init__(self, callbacks: Union[List[Callback], Callback], **kwargs):
+
+        self.callbacks = (
+            callbacks if isinstance(callbacks, list) else [callbacks]
+        )
+
+    def include(
+        self, new_callbacks: Union["CallbackList", List[Callback]]
+    ) -> None:
+        """
+        Include callbacks of a CallbackList or a list of callbacks in self.callbacks.
+        If two Callbacks are the same type, self.callbacks are prioritized and the new clalback will not be added.
+
+        Parameters
+        ----------
+        callbacks
+            A gluonts.mx.trainer.callback.CallbackList.
+        """
+
+        if not isinstance(new_callbacks, list):
+            new_callbacks = new_callbacks.callbacks
+
+        callback_types = {type(callback) for callback in self.callbacks}
+        # make sure not to have no duplicates
+        for callback in new_callbacks:
+            if type(callback) in callback_types:
```

Comment:
I wonder, why do we check that? If the API is misused, then it's not so much our problem here. Also, what if I can instantiate a class twice with different settings. This would disallow it, right?

Reply:
Yeah you're right. Let's just call it `extend` and skip the type checks.